classes/patreon_protect.php

```diff
@@ -0,0 +1,812 @@
+				$version_filename   = $attachment_metadata['sizes'][$key]['file'];
+			}
+		}
+		if ( ! isset( $cached_filename ) ) {
+			// The file was not matched in lower sizes. Then treat it as the original image
+			$cached_filename    = $attachment_id;
+			$attachment_version = 'full';
+			$version_filename   = basename( $image );
+			
+		}
+		
+		// first, check if cached image exists:
+		
+		if ( file_exists( PATREON_PLUGIN_LOCKED_IMAGE_CACHE_DIR . '/'. $cached_filename ) AND isset( $go ) ) {
+			
+			// Exists - serve the cached image:
+		
+			header( 'Content-Type: '.$mime_type );
+			// Readfile to avoid higher memory usage. Can be modified to echo file_get_contents for small files in future
+			echo file_get_contents( PATREON_PLUGIN_LOCKED_IMAGE_CACHE_DIR . '/' . $cached_filename );
+			exit;
+			
+		}	
+	
+		$image_details = wp_get_attachment_image_src( $attachment_id, $attachment_version );
+
+		$image_path = get_attached_file( $attachment_id );
+		
+		// If requested one is not the full image, replace the filename with the smaller version:
+		
+		if ( $attachment_version != 'full' ) {
+			$image_path = str_replace( $full_version_filename, $version_filename, $image_path );
+		}
+		
+		$width  = $image_details[1];
+		$height = $image_details[2];
+		
+		// Because WP does thumbnails close to 1:1 ratio, we have to check whether what's requested is a thumbnail first. So that we can serve a closer-ratio image:
+
+		// Get the thumbnail url to check against the url requested:
+		$thumbnail = wp_get_attachment_thumb_url( $attachment_id );
+		
+		if ( $thumbnail == $image OR ( $width<=150 AND $height <= 150 ) ) {
+			// Thumbnail is requested or a very small image. Override the image details with thumbnail's and double the size for better resolution in retina devices
+			
+			// If this is a very small image, it wont have a thumbnail. Check if WP created a thumbnail to use, if not, skip
+			
+			if ( isset( $attachment_metadata['sizes']['thumbnail'] ) ) {
+			
+				$image_path = str_replace( basename( $image_path ), '', $image_path ) . $attachment_metadata['sizes']['thumbnail']['file'];
+				$width      = $attachment_metadata['sizes']['thumbnail']['width'];
+				$height     = $attachment_metadata['sizes']['thumbnail']['height'];
+			}
+		}
+		
+		// For images that are very small, we wont render entire interface. So turn on some flags if the image is smaller than a certain size
+		if ( $width <= 200 OR $height <= 150 ) {
+			$hide_text = true;
+		}
+		
+		// For images that are very small, (<150) hide the button too 
+		if ( $height <= 150 ) {
+			$hide_button = true;
+		}
+		
+		if ( $mime_type == 'image/png' ) {
+			$image = imagecreatefrompng( $image_path );
+		}
+		if ( $mime_type == 'image/gif' ) {
+			$image = imagecreatefromgif( $image_path );
+		}
+		if ($mime_type =='image/jpeg') {
+			$image = imagecreatefromjpeg( $image_path );
+		}
+		if ( $mime_type == 'image/bmp' ) {
+			$image = imagecreatefrombmp( $image_path );
+		}

```

Comment:
should this be a switch statement?

Reply:
More productive to leave them as individual ifs since any of these may need modification to more complex conditionals at any time depending on changing requirements and especially server/infrastructure compatibility.